File: src/offboard_control/offboard_control/offboard_controller.py
#!/usr/bin/env python3
"""
offboard_controller.py - PX4 offboard control via DDS (px4_msgs)
Compatible with Ubuntu 22.04 + ROS2 Humble + PX4 1.14+

Subscribes to /trajectory (Path) from FUEL planner, converts to TrajectorySetpoint
Handles arming, takeoff, offboard mode switching, landing.

Frame conversions:
- ROS map: ENU (x east, y north, z up)
- PX4: NED (x north, y east, z down) and FRD body
We convert accordingly.

Usage:
ros2 run offboard_control offboard_controller --ros-args -p use_vio:=true
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from px4_msgs.msg import (
    OffboardControlMode, TrajectorySetpoint, VehicleCommand,
    VehicleLocalPosition, VehicleStatus, VehicleOdometry
)
from nav_msgs.msg import Path
from geometry_msgs.msg import PoseStamped
import math
import time

class OffboardController(Node):
    def __init__(self):
        super().__init__('offboard_controller')
        self.declare_parameter('use_vio', True)
        self.declare_parameter('takeoff_height', -1.5)  # NED, negative up
        self.declare_parameter('max_speed', 1.0)

        self.use_vio = self.get_parameter('use_vio').value
        self.takeoff_h = self.get_parameter('takeoff_height').value
        self.max_speed = self.get_parameter('max_speed').value

        # PX4 QoS
        qos_profile = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
            history=HistoryPolicy.KEEP_LAST,
            depth=1
        )

        # Publishers
        self.offboard_mode_pub = self.create_publisher(OffboardControlMode, '/fmu/in/offboard_control_mode', 10)
        self.traj_setpoint_pub = self.create_publisher(TrajectorySetpoint, '/fmu/in/trajectory_setpoint', 10)
        self.vehicle_cmd_pub = self.create_publisher(VehicleCommand, '/fmu/in/vehicle_command', 10)
        self.pose_pub = self.create_publisher(PoseStamped, '/offboard/pose', 10)

        # Subscribers
        self.local_pos_sub = self.create_subscription(VehicleLocalPosition, '/fmu/out/vehicle_local_position', self.local_pos_callback, qos_profile)
        self.status_sub = self.create_subscription(VehicleStatus, '/fmu/out/vehicle_status', self.status_callback, qos_profile)
        self.traj_sub = self.create_subscription(Path, '/trajectory', self.trajectory_callback, 10)
        self.goal_sub = self.create_subscription(PoseStamped, '/exploration/goal', self.goal_callback, 10)

        # State
        self.local_pos = None
        self.status = None
        self.current_traj = None
        self.traj_index = 0
        self.offboard_setpoint_counter = 0
        self.state = "IDLE"  # IDLE, ARMING, TAKEOFF, EXPLORING, LANDING
        self.arm_time = None

        # Timer 20Hz for offboard
        self.timer = self.create_timer(0.05, self.timer_callback)

        self.get_logger().info("Offboard Controller initialized - Waiting for PX4")

    def local_pos_callback(self, msg):
        self.local_pos = msg
        # Publish pose in map frame (ENU) for mapping
        pose = PoseStamped()
        pose.header.stamp = self.get_clock().now().to_msg()
        pose.header.frame_id = "map"
        # Convert NED to ENU: x_enu = y_ned, y_enu = x_ned, z_enu = -z_ned
        pose.pose.position.x = msg.y  # NED y -> ENU x (east)
        pose.pose.position.y = msg.x  # NED x -> ENU y (north)
        pose.pose.position.z = -msg.z
        pose.pose.orientation.w = 1.0
        self.pose_pub.publish(pose)

    def status_callback(self, msg):
        self.status = msg

    def trajectory_callback(self, msg):
        if len(msg.poses) == 0:
            return
        self.current_traj = msg.poses
        self.traj_index = 0
        self.get_logger().info(f"Received trajectory with {len(msg.poses)} waypoints")
        if self.state == "TAKEOFF":
            self.state = "EXPLORING"

    def goal_callback(self, msg):
        # Convert single goal to trajectory if no trajectory active
        if self.current_traj is None or self.traj_index >= len(self.current_traj):
            # Create simple path with one point
            path = Path()
            path.header = msg.header
            path.poses = [msg]
            self.trajectory_callback(path)

    def publish_offboard_mode(self):
        msg = OffboardControlMode()
        msg.position = True
        msg.velocity = False
        msg.acceleration = False
        msg.attitude = False
        msg.body_rate = False
        msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)
        self.offboard_mode_pub.publish(msg)

    def publish_traj_setpoint(self, x_ned, y_ned, z_ned, yaw=0.0):
        msg = TrajectorySetpoint()
        msg.position = [x_ned, y_ned, z_ned]
        msg.yaw = yaw
        msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)
        self.traj_setpoint_pub.publish(msg)

    def send_vehicle_command(self, command, param1=0.0, param2=0.0):
        msg = VehicleCommand()
        msg.command = command
        msg.param1 = param1
        msg.param2 = param2
        msg.target_system = 1
        msg.target_component = 1
        msg.source_system = 1
        msg.source_component = 1
        msg.from_external = True
        msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)
        self.vehicle_cmd_pub.publish(msg)

    def arm(self):
        self.send_vehicle_command(400, 1.0, 0.0)  # VEHICLE_CMD_COMPONENT_ARM_DISARM
        self.get_logger().info("Arming command sent")

    def takeoff(self):
        # PX4 takeoff command not used in offboard, we just publish setpoint above
        pass

    def timer_callback(self):
        self.publish_offboard_mode()

        if self.local_pos is None:
            # Need to publish setpoint before offboard
            self.publish_traj_setpoint(0.0, 0.0, self.takeoff_h)
            self.offboard_setpoint_counter += 1
            return

        # State machine
        if self.state == "IDLE":
            # Publish 10 setpoints before switching to offboard
            if self.offboard_setpoint_counter < 20:
                self.publish_traj_setpoint(0.0, 0.0, self.takeoff_h)
                self.offboard_setpoint_counter += 1
            else:
                # Switch to offboard mode
                self.send_vehicle_command(176, 1.0, 6.0)  # DO_SET_MODE, custom mode 6 = offboard
                self.get_logger().info("Offboard mode command sent")
                self.state = "ARMING"
                self.arm_time = time.time()

        elif self.state == "ARMING":
            self.publish_traj_setpoint(0.0, 0.0, self.takeoff_h)
            if time.time() - self.arm_time > 1.0:
                self.arm()
                self.state = "TAKEOFF"
                self.takeoff_start = time.time()
                self.get_logger().info("ARMING -> TAKEOFF")

        elif self.state == "TAKEOFF":
            self.publish_traj_setpoint(0.0, 0.0, self.takeoff_h)
            # Check if reached takeoff height
            if abs(self.local_pos.z - self.takeoff_h) < 0.3:
                if time.time() - self.takeoff_start > 3.0:
                    self.get_logger().info("Takeoff complete, waiting for exploration goal")
                    # Stay hovering until trajectory arrives

        elif self.state == "EXPLORING":
            if self.current_traj is None or self.traj_index >= len(self.current_traj):
                # Hover at current position
                self.publish_traj_setpoint(self.local_pos.x, self.local_pos.y, self.takeoff_h)
            else:
                # Follow trajectory
                target = self.current_traj[self.traj_index]
                # Convert ENU to NED
                # target in ENU (map): x east, y north
                # NED: x north = y_enu, y east = x_enu, z down = -z_enu
                x_ned = target.pose.position.y
                y_ned = target.pose.position.x
                z_ned = -target.pose.position.z
                # Clamp z
                z_ned = max(z_ned, self.takeoff_h - 0.5)
                z_ned = min(z_ned, -0.5)

                self.publish_traj_setpoint(x_ned, y_ned, z_ned)

                # Check if reached waypoint
                dx = self.local_pos.x - x_ned
                dy = self.local_pos.y - y_ned
                dist = math.hypot(dx, dy)
                if dist < 0.4:
                    self.traj_index += 1
                    if self.traj_index >= len(self.current_traj):
                        self.get_logger().info("Trajectory complete")

        elif self.state == "LANDING":
            self.send_vehicle_command(185, 0.0, 0.0)  # LAND

def main(args=None):
    rclpy.init(args=args)
    node = OffboardController()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/offboard_control_cpp/src/offboard_controller.cpp
#include <rclcpp/rclcpp.hpp>
#include <px4_msgs/msg/offboard_control_mode.hpp>
#include <px4_msgs/msg/trajectory_setpoint.hpp>
#include <px4_msgs/msg/vehicle_command.hpp>
#include <px4_msgs/msg/vehicle_local_position.hpp>
#include <px4_msgs/msg/vehicle_status.hpp>
#include <nav_msgs/msg/path.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <cmath>

/*
 * offboard_controller.cpp - C++ fast PX4 offboard for NIDAR
 * 20Hz control loop, ENU->NED conversion, state machine
 * Much faster than Python: no GIL, direct memory
 */

using namespace std::chrono_literals;

class OffboardController : public rclcpp::Node {
public:
  OffboardController() : Node("offboard_controller") {
    this->declare_parameter("takeoff_height", -1.5);
    takeoff_h_ = this->get_parameter("takeoff_height").as_double();

    rclcpp::QoS qos(1);
    qos.best_effort();

    offboard_pub_ = this->create_publisher<px4_msgs::msg::OffboardControlMode>("/fmu/in/offboard_control_mode", 10);
    traj_pub_ = this->create_publisher<px4_msgs::msg::TrajectorySetpoint>("/fmu/in/trajectory_setpoint", 10);
    cmd_pub_ = this->create_publisher<px4_msgs::msg::VehicleCommand>("/fmu/in/vehicle_command", 10);
    pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("/offboard/pose", 10);

    local_sub_ = this->create_subscription<px4_msgs::msg::VehicleLocalPosition>("/fmu/out/vehicle_local_position", qos, std::bind(&OffboardController::localCb, this, std::placeholders::_1));
    status_sub_ = this->create_subscription<px4_msgs::msg::VehicleStatus>("/fmu/out/vehicle_status", qos, std::bind(&OffboardController::statusCb, this, std::placeholders::_1));
    traj_sub_ = this->create_subscription<nav_msgs::msg::Path>("/trajectory", 10, std::bind(&OffboardController::trajCb, this, std::placeholders::_1));
    goal_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>("/exploration/goal", 10, std::bind(&OffboardController::goalCb, this, std::placeholders::_1));

    state_="IDLE";
    timer_ = this->create_wall_timer(50ms, std::bind(&OffboardController::timerCb, this));
    RCLCPP_INFO(this->get_logger(), "C++ Offboard Controller - takeoff %.2f NED", takeoff_h_);
  }

private:
  void localCb(const px4_msgs::msg::VehicleLocalPosition::SharedPtr msg) {
    local_=*msg; has_local_=true;
    geometry_msgs::msg::PoseStamped p;
    p.header.stamp=this->now(); p.header.frame_id="map";
    p.pose.position.x=msg->y; p.pose.position.y=msg->x; p.pose.position.z=-msg->z;
    p.pose.orientation.w=1.0;
    pose_pub_->publish(p);
  }
  void statusCb(const px4_msgs::msg::VehicleStatus::SharedPtr msg) { status_=*msg; }
  void trajCb(const nav_msgs::msg::Path::SharedPtr msg) {
    if (msg->poses.empty()) return;
    traj_=*msg; idx_=0;
    RCLCPP_INFO(this->get_logger(), "Trajectory %zu waypoints", msg->poses.size());
    if (state_=="TAKEOFF") state_="EXPLORING";
  }
  void goalCb(const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
    if (traj_.poses.empty() || idx_>=traj_.poses.size()) {
      nav_msgs::msg::Path path; path.header=msg->header; path.poses.push_back(*msg);
      traj_=path; idx_=0;
    }
  }

  void publishOffboardMode() {
    px4_msgs::msg::OffboardControlMode m;
    m.position=true; m.velocity=false; m.acceleration=false; m.attitude=false; m.body_rate=false;
    m.timestamp=this->get_clock()->now().nanoseconds()/1000;
    offboard_pub_->publish(m);
  }
  void publishSetpoint(double x_ned, double y_ned, double z_ned, double yaw=0) {
    px4_msgs::msg::TrajectorySetpoint s;
    s.position={static_cast<float>(x_ned), static_cast<float>(y_ned), static_cast<float>(z_ned)};
    s.yaw=yaw;
    s.timestamp=this->get_clock()->now().nanoseconds()/1000;
    traj_pub_->publish(s);
  }
  void sendCmd(int cmd, float p1=0, float p2=0) {
    px4_msgs::msg::VehicleCommand c;
    c.command=cmd; c.param1=p1; c.param2=p2;
    c.target_system=1; c.target_component=1; c.source_system=1; c.source_component=1; c.from_external=true;
    c.timestamp=this->get_clock()->now().nanoseconds()/1000;
    cmd_pub_->publish(c);
  }

  void timerCb() {
    publishOffboardMode();
    if (!has_local_) { publishSetpoint(0,0,takeoff_h_); counter_++; return; }

    if (state_=="IDLE") {
      if (counter_<20) { publishSetpoint(0,0,takeoff_h_); counter_++; }
      else { sendCmd(176,1,6); state_="ARMING"; arm_time_=this->now(); RCLCPP_INFO(this->get_logger(), "Offboard mode"); }
    } else if (state_=="ARMING") {
      publishSetpoint(0,0,takeoff_h_);
      if ((this->now()-arm_time_).seconds()>1.0) { sendCmd(400,1,0); state_="TAKEOFF"; takeoff_start_=this->now(); RCLCPP_INFO(this->get_logger(), "Arming -> TAKEOFF"); }
    } else if (state_=="TAKEOFF") {
      publishSetpoint(0,0,takeoff_h_);
      if (std::abs(local_.z - takeoff_h_)<0.3 && (this->now()-takeoff_start_).seconds()>3.0) {
        RCLCPP_INFO(this->get_logger(), "Takeoff complete");
      }
    } else if (state_=="EXPLORING") {
      if (traj_.poses.empty() || idx_>=traj_.poses.size()) {
        publishSetpoint(local_.x, local_.y, takeoff_h_);
      } else {
        auto &t=traj_.poses[idx_];
        double x_ned=t.pose.position.y;
        double y_ned=t.pose.position.x;
        double z_ned=-t.pose.position.z;
        z_ned=std::max(z_ned, takeoff_h_-0.5); z_ned=std::min(z_ned, -0.5);
        publishSetpoint(x_ned,y_ned,z_ned);
        double dx=local_.x-x_ned, dy=local_.y-y_ned;
        if (std::hypot(dx,dy)<0.4) { idx_++; if (idx_>=traj_.poses.size()) RCLCPP_INFO(this->get_logger(), "Traj complete"); }
      }
    }
  }

  double takeoff_h_;
  std::string state_;
  int counter_=0;
  size_t idx_=0;
  nav_msgs::msg::Path traj_;
  px4_msgs::msg::VehicleLocalPosition local_;
  px4_msgs::msg::VehicleStatus status_;
  bool has_local_=false;
  rclcpp::Time arm_time_, takeoff_start_;

  rclcpp::Publisher<px4_msgs::msg::OffboardControlMode>::SharedPtr offboard_pub_;
  rclcpp::Publisher<px4_msgs::msg::TrajectorySetpoint>::SharedPtr traj_pub_;
  rclcpp::Publisher<px4_msgs::msg::VehicleCommand>::SharedPtr cmd_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_pub_;
  rclcpp::Subscription<px4_msgs::msg::VehicleLocalPosition>::SharedPtr local_sub_;
  rclcpp::Subscription<px4_msgs::msg::VehicleStatus>::SharedPtr status_sub_;
  rclcpp::Subscription<nav_msgs::msg::Path>::SharedPtr traj_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr goal_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<OffboardController>());
  rclcpp::shutdown();
  return 0;
}
